File: Региональный этап ВСОШ 2021/2. Разбиение таблицы.py
from math import inf
for _ in range(int(input())):
    n, m = map(int, input().split())
    # Vertical razrez
    S_col = (2 + m*(n-1)) * n // 2
    l = 1
    r = m + 1
    min_verical = inf
    min_vertical_i = 0
    while l < r - 1:
        i = (l + r) // 2
        S_all_left = (2 * S_col + n*(i-2)) * (i-1) // 2
        S_col_right = ((2 * i) + m*(n-1))*n // 2
        S_all_right = ((2*S_col_right) + n*(m-i)) * (m-i+1) // 2
        # print(i, S_all_left, S_all_right)
        if S_all_left - S_all_right > 0:
            r = i
        else:
            l = i
        if abs(S_all_left - S_all_right) < min_verical:
            min_vertical_i = i
            min_verical = abs(S_all_left - S_all_right)
    # print(min_vertical_i, min_verical)

    # Horizontal razrez
    l = 1
    r = n + 1
    min_hor_i = 0
    min_hor = inf
    while l < r - 1:
        i = (l + r) // 2
        S_all_top = (1 + m*(i-1))*m*(i-1) // 2
        S_all_bottom = ((m*(i-1) + 1) + m*n) * (n*m - (m*(i-1) + 1) + 1) // 2
        if S_all_top - S_all_bottom > 0:
            r = i
        else:
            l = i
        if abs(S_all_top - S_all_bottom) < min_hor:
            min_hor_i = i
            min_hor = abs(S_all_top - S_all_bottom)
    # print(min_hor_i, min_hor)
    def f_hor(i):
        S_all_top = (1 + m*(i-1))*m*(i-1) // 2
        S_all_bottom = ((m*(i-1) + 1) + m*n) * (n*m - (m*(i-1) + 1) + 1) // 2
        return abs(S_all_top - S_all_bottom)
    def f_vert(i):
        S_all_left = (2 * S_col + n*(i-2)) * (i-1) // 2
        S_col_right = ((2 * i) + m*(n-1))*n // 2
        S_all_right = ((2*S_col_right) + n*(m-i)) * (m-i+1) // 2
        return abs(S_all_left - S_all_right)
    if min_verical <= min_hor:
        if min_vertical_i >= 3 and f_vert(min_vertical_i) == f_vert(min_vertical_i - 1):
            print("V", min_vertical_i - 1)
        else:
            print("V", min_vertical_i)
    else:
        if min_hor_i >= 3 and f_hor(min_hor_i) == f_hor(min_hor_i - 1):
            print("H", min_hor_i - 1)
        else:
            print("H", min_hor_i)

File: Региональный этап ВСОШ 2021/2. Разбиение таблицы.cpp
#include <iostream>
#include <cmath>
using namespace std;

int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    long long INF = 10e10;
    int t;
    cin >> t;
    for (int i=0; i<t; i++){
        long long n, m;
        cin >> n >> m;
        long long S_col = (2 + m*(n-1)) * n / 2;
        long long l = 0;
        long long r = m + 1;
        long long min_verical = INF;
        long long min_vertical_i = 0;
        while (l < r-1) {
            long long i = (l + r) / 2;
            long long S_all_left, S_col_right, S_all_right;
            S_all_left = (2 * S_col + n*(i-2)) * (i-1) / 2;
            S_col_right = ((2 * i) + m*(n-1))*n / 2;
            S_all_right = ((2*S_col_right) + n*(m-i)) * (m-i+1) / 2; 
            if (S_all_left - S_all_right > 0) {
                r = i;
            } else {
                l = i;
            }
            if (abs(S_all_left - S_all_right) < min_verical) {
                min_vertical_i = i;
                min_verical = abs(S_all_left - S_all_right);
            }
        }
        l = 0;
        r = n + 1;
        long long min_hor_i = 0;
        long long min_hor = INF;
        while (l < r-1) {
            long long i = (l + r) / 2;
            long long S_all_top, S_all_bottom;
            S_all_top = (1 + m*(i-1))*m*(i-1) / 2;
            S_all_bottom = ((m*(i-1) + 1) + m*n) * (n*m - (m*(i-1) + 1) + 1) / 2;
            if (S_all_top - S_all_bottom > 0) {
                r = i;
            } else {
                l = i;
            }
            if (abs(S_all_top - S_all_bottom) < min_hor) {
                min_hor_i = i;
                min_hor = abs(S_all_top - S_all_bottom);
            }
        }
        if (min_verical <= min_hor) {
            long long f_min_vert_minuse_1;
            long long i = min_vertical_i - 1;
            long long S_all_left = (2 * S_col + n*(i-2)) * (i-1) / 2;
            long long S_col_right = ((2 * i) + m*(n-1))*n / 2;
            long long S_all_right = ((2*S_col_right) + n*(m-i)) * (m-i+1) / 2;
            f_min_vert_minuse_1 = abs(S_all_left - S_all_right);
            if (min_vertical_i >= 3 and min_verical == f_min_vert_minuse_1){
                cout << "V " << min_vertical_i - 1 << "\n";
            } else {
                cout << "V " << min_vertical_i << "\n";
            }
        } else {
            long long f_min_hor_minuse_1;
            long long i = min_hor_i - 1;
            long long S_all_top = (1 + m*(i-1))*m*(i-1) / 2;
            long long S_all_bottom = ((m*(i-1) + 1) + m*n) * (n*m - (m*(i-1) + 1) + 1) / 2;
            f_min_hor_minuse_1 = abs(S_all_top - S_all_bottom);
            if (min_hor_i >= 3 and min_hor == f_min_hor_minuse_1) {
                cout << "H " << min_hor_i - 1 << "\n";
            } else {
                cout << "H " << min_hor_i << "\n";
            }
        }
    }
}
